## Problem 1



$$ 

\begin{aligned}
    min \mid A - BX \mid_{F} \\
    = tr\left( \left( A - BX \right)^{T} \left( A - BX \right)\right) \\
    = tr\left( A^T A + X^TB^T BX - X^TB^TA - A^TBX \right) \\
    = tr \left( A^T A \right) + tr \left( BXX^TB^T \right) - tr 2 \left( A^TBX \right)\\
    \\
    

    \text{Then, } \left( A^TA \right) \text{and } \left( BXX^TB^T \right) \text{where } XX^T = I  \text{ are both constant terms}.
\end{aligned}$$

Meaning that we can maximize $tr \left( A^TBX \right)$ to solve the optimization. We need to apply SVD
$$
\begin{aligned}
    tr \left( AB^TX^T \right) \\
    = tr \left(  \mathbf{U} \Sigma \mathbf{V}^T X^T \right) \\
    = tr \left(  \Sigma \mathbf{V}^T X^T \mathbf{U} \right) \text{ with } \mathbf{Z} = \mathbf{V}^T X^T \mathbf{U} \\
    = tr \left(  \Sigma \mathbf{Z} \right)
    = \sum_{i}^{} Z_{i,i}\Sigma_{i,i} \\
    
    \mathbf{Z} \text{ is an orthogonal matrix, then  the optimum is achieved when }\mathbf{Z} = \mathbf{I}
    \\
    
    \\
    
    X = \mathbf{U}\mathbf{V}^T
\end{aligned}
$$


## Problem 12

In [3]:
from run import Engine

In [4]:
engine = Engine()

### 8-point algorithm

In [4]:
engine.sfm_pipeline(use_noise=True, use_BA=True, final_vis=True, normalize_eight_point=False)

condition number 386779150735.66705
step 0, err: 7.160732705871644
step 1000, err: 0.5214589665466234
step 2000, err: 0.5633329289869181
step 3000, err: 0.5623017607094797
step 4000, err: 0.5698268473504396
step 5000, err: 0.5678945413205159
step 6000, err: 0.5560209380827632
step 7000, err: 0.5827503155670568
step 8000, err: 0.591579259857796
step 9000, err: 0.5298454106034809
[Open3D DEBUG] GLFW init.
[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display
[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display
[Open3D DEBUG] Add geometry and update bounding box to [(-22.3970, 2.1672, -28.3910) - (22.3998, 60.7769, 39.3920)]
[Open3D DEBUG] Add geometry and update bounding box to [(-22.6612, 2.0966, -28.3910) - (22.4204, 60.8290, 39.3920)]
[Open3D DEBUG] Add geometry and update bounding box to [(-22.6612, 2.0966, -28.3910) - (28.3865, 82.5838, 70.9815)]
[Open3D DEBUG] Add geometry and update bounding box to [(-22.6612, 2.0966, -28.3910) - (70.68

Condition number was huge: 386779150735.66705. The error didn't decrease much after the first 1,000 steps, with a final error of 0.5298

<img src="assets/images/bundle-adjustment-8-point.png" style="width: 500px; margin: auto; display: flex;"/>

### Normalized 8-point algorithm

In [4]:
engine.sfm_pipeline(use_noise=True, use_BA=True, final_vis=True, normalize_eight_point=True)

condition number 1331.178438294837
step 0, err: 0.39939844880027875
step 1000, err: 0.5463882982681467
step 2000, err: 0.5056460621908008
step 3000, err: 0.5696202979531187
step 4000, err: 0.512972216224318
step 5000, err: 0.5190394232732399
step 6000, err: 0.4941910075008047
step 7000, err: 0.5145344731266677
step 8000, err: 0.522201767267177
step 9000, err: 0.5102767320059329
[Open3D DEBUG] GLFW init.
[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display
[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display
[Open3D DEBUG] Add geometry and update bounding box to [(-22.3970, 2.1672, -28.3910) - (22.3998, 60.7769, 39.3920)]
[Open3D DEBUG] Add geometry and update bounding box to [(-22.3970, 1.9839, -28.3910) - (22.4352, 60.9513, 39.4400)]
[Open3D DEBUG] Add geometry and update bounding box to [(-22.3970, 1.9839, -28.3910) - (28.3865, 82.5838, 70.9815)]
[Open3D DEBUG] Add geometry and update bounding box to [(-22.3970, 1.9839, -28.3910) - (70.68

Condition number improved significantly, decreasing to 1331.178. The error stayed on similar values, with marginal improvement reaching a final error of 0.51027

<img src="assets/images/bundle-adjustment-norm-8-point.png" style="width: 500px; margin: auto; display: flex;"/>

### 2% random correspondence flipping

In [5]:
engine.sfm_pipeline(use_noise=True, use_BA=True, final_vis=True, normalize_eight_point=True, flip=True)

condition number 1328.3728785148517
step 0, err: 0.4116162877018157
step 1000, err: 0.6057451191145822
step 2000, err: 0.5955350924768572
step 3000, err: 0.5970830444593593
step 4000, err: 0.5546594614810048
step 5000, err: 0.5515708751662294
step 6000, err: 0.5379695943709637
step 7000, err: 0.5486317460873212
step 8000, err: 0.5751791939462941
step 9000, err: 0.5377793162386013
[Open3D DEBUG] GLFW init.
[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display
[Open3D WARNING] GLFW Error: Cocoa: Failed to find service port for display
[Open3D DEBUG] Add geometry and update bounding box to [(-22.3970, 2.1672, -28.3910) - (22.3998, 60.7769, 39.3920)]
[Open3D DEBUG] Add geometry and update bounding box to [(-22.3970, 2.1318, -28.3910) - (22.4384, 60.8218, 39.4583)]
[Open3D DEBUG] Add geometry and update bounding box to [(-22.3970, 2.1318, -28.3910) - (28.3865, 82.5838, 70.9815)]
[Open3D DEBUG] Add geometry and update bounding box to [(-22.3970, 2.1318, -28.3910) - (70.

Clearly we were going to receive a worse error than previously. The last step error was 0.53, worse than the previous two experiments. The condition number stayed in line with the normalized_8_point algorithm, as we keeped it enabled

<img src="assets/images/random-flipping.png" style="width: 500px; margin: auto; display: flex;"/>